pkg/backends/calico/routes.go

```diff
@@ -266,22 +390,54 @@ func (rg *routeGenerator) unsetRouteForSvc(obj interface{}) {
 	rg.Lock()
 	defer rg.Unlock()
 
-	// Remove the route if it exists.
-	if route, exists := rg.svcRouteMap[key]; exists {
-		rg.withdrawRoute(key, route)
-	}
+	routes := rg.getAdvertisedRoutes(key)
+	rg.withdrawRoutesForKey(key, routes)
+
 }
 
-// advertiseRoute advertises a route and caches it.
+// advertiseRoute advertises a route associated with the given key and
+// caches it.
 func (rg *routeGenerator) advertiseRoute(key, route string) {
-	rg.svcRouteMap[key] = route
+	if _, hasKey := rg.svcRouteMap[key]; !hasKey {
+		rg.svcRouteMap[key] = make(map[string]bool)
+	}
+
 	rg.client.AddStaticRoutes([]string{route})
+	rg.svcRouteMap[key][route] = true
+
+	rg.routeAdvertisementCount[route]++
 }
 
-// withdrawRoute withdraws a route and removes it from the cache.
+// withdrawRoute withdraws a route associated with the given key and
+// removes it from the cache.
 func (rg *routeGenerator) withdrawRoute(key, route string) {
-	rg.client.DeleteStaticRoutes([]string{route})
-	delete(rg.svcRouteMap, key)
+	// Only remove the advertisement if there are no other reasons this route
+	// should be advertised. For example, k8s will allow you to manually assign
+	// the same External IP to two different services; assign an External IP to
+	// a service which is the same as the service's cluster IP,
+	// and assign the same External IP twice to a service. In all of these
+	// scenarios, you would end up in a situation where the same route is
+	// "legitimately" being advertised twice from a node.
+	if rg.routeAdvertisementCount[route] == 1 {
+		rg.client.DeleteStaticRoutes([]string{route})
+	}
+	rg.routeAdvertisementCount[route]--
```

Comment:
I think we want to delete the route key here if the count goes to 0 so we don't leak keys in the map.